### PR TITLE
Add functionality for favorite movies -> Get favorite movies from database

### LINKED ISSUE #132: Add functionality for favorite movies
## Description
I want to add a feature of adding a movie to my favorite.

## What I should add:
- Placeholder if the list will be empty
- I will add new value to the movie data class flag `isFavorite`
- I will use update method on room database
- I should consider use flow if updating has any problem
- Icon heart (favorite) to each element in the list of movies
- Method onClick updating movie to favorite or removing from favorite
- Consider the case when the User adds a movie to his favorite and he hasn't fetched details for the movie what am I supposed to do with it? I should always try to fetch data or only when the user clicks on the movie on the list of movies or favorite movies.   

- [x] Add heart icon to movie list
- [x] Add favorite filed in popular movies
- [x] Scheme to change color if movie is favorite 
- [x] Update isFavorite filed in database
- [x] Get favorite movies from database
- [ ] Create UI for favourite movies

### PR DESCRIPTION
**Pull request by Nataniel-Antosik**

### CODE REVIEW

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt [hunk 1]
```diff
@@ -27,4 +29,8 @@ internal object MoviesModule {
     @Provides
     fun provideMovieDetailsRepository(apiService: MoviesApi, database: MovieDatabase): MovieDetailsRepository =
         MovieDetailsDataRepository(apiService, database.dao)
+
+    @Provides
+    fun provideFavouriteMoviesRepository(database: MovieDatabase): FavouriteMoviesRepository =
```
Comment (marcinWisnia) [inline data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt:34]: Let's decide and use one version of `favorite` word :D i see that in few places different versions are used :D

Comment (marcinWisnia) [inline data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: Let's decide and use one version of favorite word :D i see that in few places different versions are used :D

File: data/src/main/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepository.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import com.antosik.benchproject.data.movies.common.database.dao.MovieDao
+import com.antosik.benchproject.data.popular.movies.entity.database.toDomain
+import com.antosik.benchproject.domain.favourite.repo.FavouriteMoviesRepository
+import kotlinx.coroutines.flow.map
+
+internal class FavouriteMoviesDataRepository(
+    private val dao: MovieDao,
+) : FavouriteMoviesRepository {
+
+    override fun getFavouriteMovies() =
```
Comment (marcinWisnia) [inline data/src/main/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepository.kt:13]: i think we can omit get prefix

File: data/src/main/kotlin/com/antosik/benchproject/data/movies/common/database/dao/MovieDao.kt [hunk 3]
```diff
@@ -11,13 +11,17 @@ import com.antosik.benchproject.data.movie.details.entity.database.MovieDetailsE
...
+import kotlinx.coroutines.flow.Flow
 
 @Dao
 internal interface MovieDao {
 
     @Query("SELECT * FROM MovieEntity")
     suspend fun getMovies(): List<MovieEntity>
 
+    @Query("SELECT * FROM MovieEntity WHERE isFavorite = 1")
+    fun getFavoriteMovies(): Flow<List<MovieEntity>>
```
Comment (marcinWisnia) [inline data/src/main/kotlin/com/antosik/benchproject/data/movies/common/database/dao/MovieDao.kt]: here also, but it is nothing serious for me :D just favouriteMovies will be enough, usage of java style get can be omited

File: data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        mockkStatic(List<MovieEntity>::toDomain)
+    }
+
+    @AfterEach
+    fun afterEach() {
+        unmockkAll()
+    }
+
+    @TestFactory
+    fun `SHOULD return expected favorite movies WHEN specific data is present in the database`() =
```
Comment (marcinWisnia) [inline data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt]: do we really need a dynamic test here ? do we have any logic that is different when the list is empty or not ?

File: domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.jupiter.api.DynamicTest
+import org.junit.jupiter.api.TestFactory
+
+internal class GetFavouriteMoviesUseCaseTest {
+
+    val favouriteMoviesRepository: FavouriteMoviesRepository = mockk()
+    val tested = GetFavouriteMoviesUseCase(favouriteMoviesRepository)
+
+    @TestFactory
+    fun `SHOULD return expected favorite movies WHEN specific data incoming from repository`() =
```
Comment (marcinWisnia) [inline domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt]: do we really need a dynamic test here ? do we have any logic that is different when the list is empty or not ?

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt [hunk 6]
```diff
@@ -27,4 +29,8 @@ internal object MoviesModule {
     @Provides
     fun provideMovieDetailsRepository(apiService: MoviesApi, database: MovieDatabase): MovieDetailsRepository =
         MovieDetailsDataRepository(apiService, database.dao)
+
+    @Provides
```
Comment (marcinWisnia) [inline data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: Alternatively, we could do the repo impl. injectable and use @Binds annotation :) please check this option :D

**Review by marcinWisnia — CHANGES_REQUESTED** [review]

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt (hunk 6 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: I used it some months ago for example here https://github.com/Nataniel-Antosik/Bench_Project/pull/10/files

```
@Module
@InstallIn(ViewModelComponent::class)
interface RepositoryModule {

    @Binds
    fun bindPopularMoviesRepository(impl: PopularMoviesDataRepository): PopularMoviesRepository
}
```

```
@Module
@InstallIn(SingletonComponent::class)
object MoviesModule {

...

    @Singleton
    @Provides
    fun providePopularMoviesDataRepository(apiService: MoviesApi) = PopularMoviesDataRepository(apiService)
}
```
And I remember the reason why I don't use `binds` is provides are shorter than creating binds and provider

File: data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt (hunk 4 above)
Comment (Nataniel-Antosik) [reply data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt]: To be honest we didn't have any logic when the list is empty or not. I would like just to test both cases but if you think one test will be enough I can go with it.

File: domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt (hunk 5 above)
Comment (Nataniel-Antosik) [reply domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt]: To be honest we didn't have any logic when the list is empty or not. I would like just to test both cases but if you think one test will be enough I can go with it.

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt (hunk 1 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: Ok, I will choose one version :P

File: data/src/main/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepository.kt (hunk 2 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepository.kt:13]: In the whole app, I have this prefix, I can change it but what about when I would like to update something it is ok to use the prefix `update`?

File: data/src/main/kotlin/com/antosik/benchproject/data/movies/common/database/dao/MovieDao.kt (hunk 3 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/movies/common/database/dao/MovieDao.kt]: Same like here 😄 https://github.com/Nataniel-Antosik/Bench_Project/pull/182#discussion_r1284193272

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt (hunk 1 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: Done here: #183 😄

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt (hunk 6 above)
Comment (marcinWisnia) [reply data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: hmm, what do You mean by creating bind and provider ? repo impl could be done with @Inject, then we just need one line with binds in the module, or i'm missing something :D ?

File: data/src/main/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepository.kt (hunk 2 above)
Comment (marcinWisnia) [reply data/src/main/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepository.kt:13]: fully ok :D

File: data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt (hunk 4 above)
Comment (marcinWisnia) [reply data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt]: yes but in my opinion it is useless, as long as we don't have any logic behind if the list is empty :) there is no condition for that or handling this as returning specified sealed class object because of handling some specific situation in viewmodel or something :D so in this case i would really drop those ones :D event if it is some logic during the mapping, the mapper test could resolve those cases at all :D

File: domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt (hunk 5 above)
Comment (marcinWisnia) [reply domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt]: same as in previous comment :)

**Review by marcinWisnia — CHANGES_REQUESTED** [review]

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt (hunk 6 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: After our discussion I created two issues #184, #185 😄

File: data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt (hunk 4 above)
Comment (Nataniel-Antosik) [reply data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt]: Done 😄

File: domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt (hunk 5 above)
Comment (Nataniel-Antosik) [reply domain/src/test/kotlin/com/antosik/benchproject/domain/favourite/usecase/GetFavouriteMoviesUseCaseTest.kt]: Done 😄

File: data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt (hunk 6 above)
Comment (Nataniel-Antosik) [reply data/src/main/kotlin/com/antosik/benchproject/data/common/modules/movies/MoviesModule.kt]: Fixed here #186

File: data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @AfterEach
+    fun afterEach() {
+        unmockkAll()
+    }
+
+    @Test
+    fun `SHOULD return favourite movies WHEN data is present in the database`() = runTest {
+        val movies = listOf(mockk<MovieModel>())
```
Comment (marcinWisnia) [inline data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt]: minor: this just can be a mocked list if possible :)

**Review by marcinWisnia — APPROVED** [review]

File: data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt (hunk 7 above)
Comment (Nataniel-Antosik) [reply data/src/test/kotlin/com/antosik/benchproject/data/favourite/repo/FavouriteMoviesDataRepositoryTest.kt]: Done 😄
